### PR TITLE
Import video transcripts

### PR DESCRIPTION
**Pull request by mushtaqak**
Related Platform PR: https://github.com/edx/edx-platform/pull/17589

[EDUCATOR-2173](https://openedx.atlassian.net/browse/EDUCATOR-2173)

[Sandbox URL](https://studio-importvideotranscripts.sandbox.edx.org/import/course-v1:edX+DemoX+Demo_Course)

### CODE REVIEW

**Comment by mushtaqak** [comment]
@muhammad-ammar @Qubad786 Review?

File: edxval/tests/test_api.py [hunk 1]
```diff
@@ -1150,29 +1169,32 @@ def assert_encoded_video_matches_dict(self, encoded_video, encoded_video_dict):
     def assert_invalid_import(self, xml, course_id=None):
         edx_video_id = "test_edx_video_id"
         with self.assertRaises(ValCannotCreateError):
-            api.import_from_xml(xml, edx_video_id, course_id)
+            api.import_from_xml(xml, edx_video_id, self.file_system, course_id)
         self.assertFalse(Video.objects.filter(edx_video_id=edx_video_id).exists())
 
     def assert_transcripts(self, video_id, expected_transcripts):
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: We should also assert transcript contents.

File: edxval/api.py [hunk 2]
```diff
@@ -968,26 +964,44 @@ def import_from_xml(xml, edx_video_id, course_id=None):
...
     Arguments:
-        xml (Element): lxml Element object
+        xml (Element): lxml Element object.
+        edx_video_id (str): Video id of the video.
+        resource_fs (OSFS): Import file system.
     """
     for transcript in xml.findall('.//transcripts/transcript'):
         try:
-            create_or_update_video_transcript(
-                transcript.attrib['video_id'],
```
[the remark this answers is not in the document]
Comment (mushtaqak) [reply edxval/api.py]: We have decided to not to include `video_id` in transcript tag

File: edxval/api.py [hunk 3]
```diff
@@ -968,26 +964,44 @@ def import_from_xml(xml, edx_video_id, course_id=None):
             'bitrate': encoded_video_el.get('bitrate'),
         })
     create_video(data)
-    create_transcript_objects(xml)
+    create_transcript_objects(xml, edx_video_id, resource_fs)
```
[the remark this answers is not in the document]
Comment (mushtaqak) [reply edxval/api.py]: We have decided to not to include video_id in transcript tag

File: edxval/api.py [hunk 4]
```diff
@@ -915,7 +911,7 @@ def import_from_xml(xml, edx_video_id, course_id=None):
     if xml.tag != 'video_asset':
         raise ValCannotCreateError('Invalid XML')
 
-    # TODO this will be moved as a part of EDUCATOR-2173
+    # TODO this will be moved as a part of EDUCATOR-2403
```
[the remark this answers is not in the document]
Comment (mushtaqak) [reply edxval/api.py]: This is the case for external videos which is going to be addressed in EDUCATOR-2403

**Comment by mushtaqak** [comment]
@Qubad786 Addressed your feedback. 

@muhammad-ammar need your review on this as well

**Comment by muhammad-ammar** [comment]
@mushtaqak I am going to start the review now.

File: edxval/tests/test_views.py [hunk 5]
```diff
@@ -813,6 +813,7 @@ def test_create_transcript(self):
         Tests POSTing transcript successfully.
         """
         post_transcript_data = dict(self.transcript_data)
+        post_transcript_data.pop('file_data')
```
Comment (muhammad-ammar) [inline edxval/tests/test_views.py]: why this change?

File: edxval/tests/test_api.py [hunk 6]
```diff
@@ -1333,37 +1380,158 @@ def test_invalid_course_id(self):
...
+        self.assert_transcripts(
+            constants.VIDEO_DICT_FISH['edx_video_id'],
+            []
+        )
+
+    def test_import_transcript_new_video(self):
+        """
+        Verify that transcript import for new video is working as expected when transcript is present in XML.
+        """
+        new_course_id = 'new_course_id'
```
Comment (muhammad-ammar) [inline edxval/tests/test_api.py]: unused variable

File: edxval/tests/test_api.py [hunk 7]
```diff
@@ -1150,29 +1170,32 @@ def assert_encoded_video_matches_dict(self, encoded_video, encoded_video_dict):
...
             ).data
-            received['name'] = received.pop('url')
 
-            self.assertDictEqual(received, expected_transcript)
+            # Remove not needed keys.
+            expected_transcript.pop('transcript', None)
+            expected_transcript.pop('file_data', None)
+            received_transcript.pop('url')
+
+            self.assertDictEqual(received_transcript, expected_transcript)
```
Comment (muhammad-ammar) [inline edxval/tests/test_api.py]: don't we need to compare the transcript content?

File: edxval/tests/test_api.py [hunk 8]
```diff
@@ -1333,37 +1380,158 @@ def test_invalid_course_id(self):
...
+            self.file_system,
+            constants.EXPORT_IMPORT_STATIC_DIR,
+            'test_course_id'
+        )
+
+        # Verify that no transcript record is created.
+        self.assert_transcripts(
+            constants.VIDEO_DICT_FISH['edx_video_id'],
+            []
+        )
```
Comment (muhammad-ammar) [inline edxval/tests/test_api.py]: why about using the below instead of `assert_transcripts`?
```python
with self.assertRaises(VideoTranscript.DoesNotExist):
    VideoTranscript.objects.get(video__edx_video_id=constants.VIDEO_DICT_FISH['edx_video_id'])
```

File: edxval/tests/test_api.py [hunk 9]
```diff
@@ -1333,37 +1380,158 @@ def test_invalid_course_id(self):
...
+            '',
+            self.file_system,
+            constants.EXPORT_IMPORT_STATIC_DIR
+        )
         self.assertEqual(
             VideoTranscript.objects.count(),
             0
         )
 
+    def test_import_transcript_attached_existing_video(self):
```
Comment (muhammad-ammar) [inline edxval/tests/test_api.py]: we also need a test which verify the below
```markdown
1. we have video with `edx_video_id=12345` with some transcripts
3. do import with new transcripts
4. verify that video with `edx_video_id=12345` has both the old and new transcripts
```

File: edxval/tests/test_views.py (hunk 5 above)
Comment (mushtaqak) [reply edxval/tests/test_views.py]: `post_transcript_data` has `file_data` which we added to use for content. This isn't needed in this test. This is already happening in case of `video_id` as well.

File: edxval/tests/test_api.py (hunk 9 above)
Comment (mushtaqak) [reply edxval/tests/test_api.py]: If we have video with some edx_video_id in DS then we don't do import for it (therefore transcript import is also not done)

**Comment by mushtaqak** [comment]
@muhammad-ammar addressed your feedback

File: edxval/api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Qubad786) [reply edxval/api.py]: sounds good.

File: edxval/api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Qubad786) [reply edxval/api.py]: Oh I forgot!

File: edxval/utils.py [hunk 10]
```diff
@@ -169,3 +171,17 @@ def get_video_transcript_storage():
         # during edx-platform loading this method gets called but settings are not ready yet
         # so in that case we will return default(FileSystemStorage) storage class instance
         return get_storage_class()()
+
+
+def create_file_to_fs(file_data, file_name, file_system, static_dir):
```
Comment (Qubad786) [inline edxval/utils.py]: `write_file_to_fs` or `create_file_in_fs `.

File: edxval/tests/test_api.py [hunk 11]
```diff
@@ -939,14 +925,16 @@ def setUp(self):
         # create internal video transcripts
         transcript_data = dict(constants.VIDEO_TRANSCRIPT_CIELO24, video=video)
         transcript_data.pop('video_id')
+        transcript_data.pop('file_data')
         VideoTranscript.objects.create(**transcript_data)
         transcript_data = dict(constants.VIDEO_TRANSCRIPT_3PLAY, video=video)
         transcript_data.pop('video_id')
+        transcript_data.pop('file_data')
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: we can create a helper method to omit un-needed attributes, the above three lines would become one.
```bash
def omit_attrs(dict, attrs_to_omit=[]):
    return {attr: value for attr, value in dict.iteritems() if attr not in attrs_to_omit}
```

File: edxval/tests/test_api.py [hunk 12]
```diff
@@ -1150,29 +1170,37 @@ def assert_encoded_video_matches_dict(self, encoded_video, encoded_video_dict):
...
 
-            self.assertDictEqual(received, expected_transcript)
+            # Assert transcript content
+            expected_transcript_content = expected_transcript.get('file_data', None)
+            received_transcript['file_data'] = json.loads(
+                api.get_video_transcript_data(video_id, language_code)['content']
+            )
+
+            # Remove not needed keys.
+            expected_transcript.pop('transcript', None)
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: we can use: `assertDictContainsSubset` instead of popping items.

Comment (mushtaqak) [reply edxval/tests/test_api.py]: actually `recieved_transcript.url` is saved with uuid so we can't check with `expected_transcript.url` which has static file url.

**Comment by mushtaqak** [comment]
@muhammad-ammar Need your thumbs up

File: edxval/api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Qubad786) [reply edxval/api.py]: @mushtaqak is this addressed yet?

File: edxval/api.py [hunk 13]
```diff
@@ -968,26 +964,43 @@ def import_from_xml(xml, edx_video_id, course_id=None):
...
-                metadata=dict(
-                    provider=transcript.attrib['provider'],
-                    file_name=transcript.attrib['file_name'],
-                    file_format=transcript.attrib['file_format'],
+            file_format = transcript.attrib['file_format']
+            language_code = transcript.attrib['language_code']
+            transcript_data = get_video_transcript_data(edx_video_id, language_code)
+
+            # First check if transcript record does not exist.
+            if not transcript_data:
```
Comment (Qubad786) [inline edxval/api.py]: When will this exist? Is this a valid scenario ? because I guess if a video is already there then we dont do any further import.

File: edxval/tests/test_api.py [hunk 14]
```diff
@@ -1150,29 +1175,37 @@ def assert_encoded_video_matches_dict(self, encoded_video, encoded_video_dict):
...
-            received = api.TranscriptSerializer(
+            # Get the imported transcript and rename `url` key.
+            received_transcript = api.TranscriptSerializer(
                 VideoTranscript.objects.get(video__edx_video_id=video_id, language_code=language_code)
             ).data
-            received['name'] = received.pop('url')
 
-            self.assertDictEqual(received, expected_transcript)
+            # Assert transcript content
+            expected_transcript_content = expected_transcript.get('file_data', None)
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: nit: `expected_transcript['file_data']`

File: edxval/tests/test_api.py [hunk 15]
```diff
@@ -1333,37 +1388,152 @@ def test_invalid_course_id(self):
...
+        )
+
+        # Verify video is present before.
+        video = Video.objects.get(edx_video_id=constants.VIDEO_DICT_FISH['edx_video_id'])
+        self.assertIsNotNone(video)
+
+        # Create internal video transcripts
+        transcript_data = dict(constants.VIDEO_TRANSCRIPT_3PLAY, video=video)
+        transcript_data.pop('video_id')
+        transcript_data.pop('file_data')
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: These can be: `transcript_data  = omit_attrs(transcript_data, ['video_id', 'file_data'])`

File: edxval/tests/test_api.py [hunk 16]
```diff
@@ -1333,37 +1388,152 @@ def test_invalid_course_id(self):
...
         """
         video_id = 'super-soaker'
-        transcript_xml = '<transcript file_name="wow.srt" language_code="en" file_format="srt" provider="Cielo24"/>'
+        transcript_xml = '<transcript file_format="srt" provider="Cielo24"/>'
         xml = etree.fromstring("""
             <video_asset>
                 <transcripts>
                     {transcript_xml}
-                    <transcript file_name="edxval/tests/data/wow.sjson" language_code="de" file_format="sjson" provider='3PlayMedia' video_id="{video_id}"/>
+                    <transcript language_code="de" file_format="sjson" provider='3PlayMedia' video_id="{video_id}"/>
```
Comment (Qubad786) [inline edxval/tests/test_api.py]: I think we don't export `video_id` right?

File: edxval/api.py (hunk 13 above)
Comment (mushtaqak) [reply edxval/api.py]: Yes, this is a case when we have transcripts already stored with the same language.

File: edxval/api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mushtaqak) [reply edxval/api.py]: Yes in #133

File: edxval/api.py (hunk 13 above)
Comment (Qubad786) [reply edxval/api.py]: Got it!
